src/app/utils/field-validation/validators/tableValidator.ts

```diff
@@ -0,0 +1,146 @@
+type TableValidatorConstructor = (
+  tableField: ITableFieldSchema,
+) => TableValidator
+
+/**
+ * Returns a validation function to check if the
+ * response has less than the minimum number of rows specified.
+ */
+const makeMinimumRowsValidator: TableValidatorConstructor = (tableField) => (
+  response,
+) => {
+  const { answerArray } = response
+  const { minimumRows } = tableField
+
+  return answerArray.length >= minimumRows
+    ? right(response)
+    : left(`TableValidator:\tanswer has less than the minimum number of rows`)
+}
+
+/**
+ * Returns a validation function to check if addMoreRows is not set
+ * and if so, whether the response has more than minimum number of rows.
+ */
+const makeAddMoreRowsValidator: TableValidatorConstructor = (tableField) => (
+  response,
+) => {
+  const { answerArray } = response
+  const { minimumRows, addMoreRows } = tableField
+
+  if (addMoreRows) return right(response)
+  return answerArray.length === minimumRows
+    ? right(response)
+    : left(
+        `TableValidator:\tanswer has extra rows even though addMoreRows is false`,
+      )
+}
+
+/**
+ * Returns a validation function to check if the
+ * response has more than the maximum number of rows specified.
+ */
+const makeMaximumRowsValidator: TableValidatorConstructor = (tableField) => (
+  response,
+) => {
+  const { answerArray } = response
+  const { maximumRows } = tableField
+
+  if (!maximumRows) return right(response)
+
+  return answerArray.length <= maximumRows
+    ? right(response)
+    : left(`TableValidator:\tanswer has more than the maximum number of rows`)
+}
+
+/**
+ * Returns a validation function to check if the
+ * response has the correct number of answers for each row.
+ */
+const makeRowLengthValidator: TableValidatorConstructor = (tableField) => (
+  response,
+) => {
+  const { answerArray } = response
+  const { columns } = tableField
+
+  return answerArray.every((row) => row.length === columns.length)
+    ? right(response)
+    : left(`TableValidator:\tanswer has rows with incorrect number of answers`)
+}
+
+/**
+ * Returns a validation function that checks that the columns only have allowed types
+ */
+
+const makeColumnTypeValidator: TableValidatorConstructor = (tableField) => (
+  response,
+) => {
+  const { columns } = tableField
+  return columns.every((column) =>
+    ALLOWED_COLUMN_TYPES.includes(column.columnType),
+  )
+    ? right(response)
+    : left(`TableValidator:\tanswer has columns with non-allowed types`)
+}
+
+/**
+ * Returns a validation function that applies
+ * the correct validator for each column based on its type.
+ */
+
+const makeColumnValidator: TableValidatorConstructor = (tableField) => (
+  response,
+) => {
+  const { answerArray, isVisible, _id } = response
+  const { columns } = tableField
+
+  return answerArray.every((row) => {
+    return row.every((answer, i) => {
+      const answerField = createAnswerFieldFromColumn(columns[i])
```

Comment:
Notice that the performance of this algorithm is sub-optimal because `answerField` is re-constructed over and over again.

How about using _function composition_ to construct a custom `validateRow` function out of the `columns` variable first, before applying it to each row? I think you already have all the constituent parts to do so.

Reply:
See if the revised implementation works, which calls createAnswerFieldFromColumn() only once for each column (instead of once for each cell previously):
```
  const answerFieldColumns = columns.map((column) =>
    createAnswerFieldFromColumn(column),
  )

  return answerArray.every((row) => {
    return row.every((answer, i) => {
      const answerField = answerFieldColumns[i]
```